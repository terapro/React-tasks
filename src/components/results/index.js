import React, {Component} from 'react';

import PropTypes from 'prop-types';

import {NoResults} from 'src/components/results/no-results/no-results.js';
import {Item} from 'src/components/results/item/item.js';
import {Summary} from 'src/components/summary/index.js'

class Results extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sortItems: {
        list: ['release date', 'rating'],
        active: 'release date'
      }
    };
    /**
     * // CallBack for changing this.state (it changes active genre)
     * @param {element} el
     */
    this.changeSortItemClick = (el) => {
      this.setState({sortItems: {
          list: this.state.sortItems.list,
          active: el.target.id
        }})
    };
    /**
     * // Function for sorting results
     * @param {Object} res
     * @param {string} parameter
     * @returns {DependencyReference[]|void|*}
     */
    this.sortResult = (res, parameter) => {
      const sortParametersMatching = {
        'release date': 'release_date',
        'rating': 'vote_average'
      };

      parameter = sortParametersMatching[parameter];

      switch (parameter) {
        case 'release_date' : {
          return res.sort(sortFuncDates);
        }
        case 'vote_average': {
          return res.sort(sortFuncRating);
        }
      }

      /**
       * Sort function for Rating
       * @param {Object} a
       * @param {Object} b
       * @returns {number}
       */
      function sortFuncRating (a,b) {
        return +b[parameter] - +a[parameter]
      }

      /**
       *Sort function for Release Date
       * @param {Object} a
       * @param {Object} b
       * @returns {number}
       */
      function sortFuncDates(a,b) {
        return Date.parse(b[parameter]) - Date.parse(a[parameter]);
      }
    }
  }
  render() {
    let {searchResult, setFilmModeCallback, filmMode, filmModeGenre, searchPhrase} = this.props;
   // if (filmMode) {throw new Error('Film mode')}
    searchResult = this.sortResult(searchResult, this.state.sortItems.active); // Sort result here
    if (searchResult.length) { // When at least 1 film to show
      return (
        <React.Fragment>
          <Summary filmNumber={searchResult.length}
                   sortItems={this.state.sortItems}
                   changeSortItemCallBack={this.changeSortItemClick}
                   filmMode={filmMode}
                   filmModeGenre={filmModeGenre}
                   searchPhrase={searchPhrase}
          />
          <div className='results'>
            {
              searchResult.map((item) => (
                                          <Item key = {item['id']} info = {item} setFilmModeCallback={setFilmModeCallback} />
                                        )
                          )
            }
          </div>
        </React.Fragment>
      );
    } else {
      return (
        <div className='results'>
          <NoResults />
        </div>

      );
    }
  }
}

Results.propTypes = {
  searchPhrase: PropTypes.string,
  filmMode: PropTypes.bool,
  filmModeGenre: PropTypes.string,
  searchResult: PropTypes.array,
  setFilmModeCallback: PropTypes.func
};

Results.defaultProps = {
  searchResult: []
};

export {Results};