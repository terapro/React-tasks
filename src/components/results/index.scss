.results {
  min-height: 110px;
  width: inherit;
  background: #fefefe;
  text-align: center;
  font-size: 40px;
  padding-top: 200px;
}