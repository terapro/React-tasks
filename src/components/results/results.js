import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import {openFilm} from "src/actions";
import {NoResults} from 'src/components/results/no-results/no-results.js';
import {Item} from 'src/components/results/item/item.js';
import {Summary} from 'src/components/summary/summary.js';


class ResultsChild extends Component {
  constructor(props) {
        super(props);
        /**
         * Sorts the list by the sorting parameters
         * @param {array} res - array of film objects
         * @param parameter
         * @returns {DependencyReference[]|void|*}
         */
    this.sortResult = (res, parameter) => {
      const sortParametersMatching = {
        'release date': 'release_date',
        'rating': 'vote_average'
      };
      parameter = sortParametersMatching[parameter];

      switch (parameter) {
        case 'release_date' : {
          return res.sort(sortFuncDates);
          }
          case 'vote_average': {
            return res.sort(sortFuncRating);
          }
      }
            /**
             * Sort function for Rating
             * @param {Object} a
             * @param {Object} b
             * @returns {number}
             */
            function sortFuncRating(a, b) {
                return +b[parameter] - +a[parameter]
            }

            /**
             *Sort function for Release Date
             * @param {Object} a
             * @param {Object} b
             * @returns {number}
             */
            function sortFuncDates(a, b) {
                return Date.parse(b[parameter]) - Date.parse(a[parameter]);
            }
        };
    this.filmList=[];
    this.chooseFilmByClick = (el)=> {
        const index = el.target.id;
        const film = this.filmList[index];
        this.props.onOpenFilm(film);
    };
  }
  render() {
    const {searchList, recommendedList, filmMode, sortBy} = this.props;
      this.filmList = filmMode? [...recommendedList] : [... searchList]; // The film list depends on mode: search/film
      this.filmList = this.sortResult(this.filmList, sortBy.chosenParameter);

    if (this.filmList.length) { // When at least 1 film to show
      return (
        <Fragment>
          <Summary />
          <div className='results'>
            {this.filmList.map((item, i) => (
                            <Item key={item['id']}
                                  info={item}
                                  onPosterClick={this.chooseFilmByClick}
                                  filmIndex={i}// we store the array index in the key property to save the quick reference to the film object when clicking the poster
                            />
            ))}
          </div>
        </Fragment>
      );
    } else {
      return (
        <div className='results'>
          <NoResults />
        </div>
      );
    }
  }
}

ResultsChild.propTypes = {
    searchList: PropTypes.array,
    recommendedList: PropTypes.array,
    onOpenFilm: PropTypes.func,
    onChangeItem: PropTypes.func,
    filmMode: PropTypes.bool,
    sortBy: PropTypes.shape({
        parameters: PropTypes.arrayOf(PropTypes.string),
        chosenParameter: PropTypes.string
    }).isRequired
};
ResultsChild.defaultProps = {
    searchList: [],
    recommendedList: [],
    onOpenFilm: f=>f,
    filmMode: false,
    onChangeItem: f=>f
};

export const Results = connect(
  store =>
    ({
      searchAttributesFromStore: store.search,
      searchList: store.films.searchList,
      recommendedList: store.films.recommendedList,
      filmMode: store.mode.film,
      sortBy: store.films.sortBy
    }),
  dispatch =>
    ({
      onOpenFilm(film) {
        dispatch(openFilm(film))
      }
    })
)(ResultsChild);


