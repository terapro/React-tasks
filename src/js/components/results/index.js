import React from 'react';

const Results = () => (
    <div className={'results'}>
        No films found
    </div>);


export {Results};