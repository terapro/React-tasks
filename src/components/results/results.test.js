import React from 'react';
import {Results} from 'src/components/results/results.js';
import {shallow} from 'enzyme';

const mockFunc = jest.fn();

const mockNoResults = [];

const mockProps = {
  searchPhrase: 'Searching something',
  filmModeGenre: 'action',
  setFilmMode: jest.fn(),
  filmMode: false,
  searchInFilmModeByGenre: jest.fn()
};

const mockEl ={target: {id: 'rating'}};

const mockFilm = {
  "id": 333339,
  "title": "Ready Player One",
  "tagline": "A better reality awaits.",
  "vote_average": 8.1,
  "vote_count": 617,
  "release_date": "2018-03-28",
  "poster_path": "https://image.tmdb.org/t/p/w500/pU1ULUq8D3iRxl1fdX2lZIzdHuI.jpg",
  "overview": "When the creator of a popular video game system dies, a virtual contest is created to compete for his fortune.",
  "budget": 175000000,
  "revenue": 0,
  "genres": [
    "Adventure",
    "Science Fiction",
    "Action"
  ],
  "runtime": 140
};
const mockSomeResults = [mockFilm];

describe('<Results />', () => {
  describe('Rendering', () => {
    it('should render the component', () => {
      const wrapper = shallow(
        <Results searchInFilmModeByGenre={mockProps.searchInFilmModeByGenre}
                 searchResult={mockSomeResults}
                 setFilmMode={mockProps.setFilmMode}
                 filmModeGenre={mockProps.filmModeGenre}
                 filmMode={mockProps.filmMode}
                 searchPhrase={mockProps.searchPhrase}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
    it('should render the component without results', () => {
      const wrapper = shallow(
        <Results searchInFilmModeByGenre={mockProps.searchInFilmModeByGenre}
                 searchResult={mockNoResults}
                 setFilmMode={mockProps.setFilmMode}
                 filmModeGenre={mockProps.filmModeGenre}
                 filmMode={mockProps.filmMode}
                 searchPhrase={mockProps.searchPhrase}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('Methods', () => {
    let instance;
    beforeAll(() => {
      instance = shallow(<Results searchInFilmModeByGenre={mockProps.searchInFilmModeByGenre}
                                   searchResult={mockSomeResults}
                                   setFilmMode={mockProps.setFilmMode}
                                   filmModeGenre={mockProps.filmModeGenre}
                                   filmMode={mockProps.filmMode}
                                   searchPhrase={mockProps.searchPhrase}
      />).instance();
    });
    afterEach(() => {
      jest.clearAllMocks();
    });
    describe('changeSortItemClick', () => {
      it('should set the active sort item in the state', () => {
        instance.state = {
          sortItems: {
            list: ['release date', 'rating'],
            active: 'release date'
          }
        };
        instance.changeSortItemClick(mockEl);
        expect(instance.state.sortItems.active).toEqual(mockEl.target.id);

      });
    });
  });
});



  /*describe('Methods', () => {
    let instance;
    beforeAll(() => {
      instance = shallow(
        <Results searchInFilmModeByGenre={mockProps.searchInFilmModeByGenre}
                 searchResult={{}}
                 setFilmMode={mockProps.setFilmMode}
                 filmModeGenre={mockProps.filmModeGenre}
                 filmMode={mockProps.filmMode}
                 searchPhrase={mockProps.searchPhrase}
        />
      ).instance();
    });
    afterEach(() => {
      jest.clearAllMocks();
    });
    it('should change the sort item', () => {
      instance.state = {
        sortItems: {
          list: ['release date', 'rating'],
          active: 'release date'
        }
      };
      instance.changeSortItemClick(mockEl);
      expect(instance.state.sortItems.active).toEqual(mockEl.target.id);
    });

  });*/


