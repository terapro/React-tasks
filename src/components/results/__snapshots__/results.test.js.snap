// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Results /> Rendering should render the component 1`] = `
<React.Fragment>
  <Summary
    changeSortItemCallBack={[Function]}
    filmMode={false}
    filmModeGenre="action"
    filmNumber={1}
    searchPhrase="Searching something"
    sortItems={
      Object {
        "active": "release date",
        "list": Array [
          "release date",
          "rating",
        ],
      }
    }
  />
  <div
    className="results"
  >
    <Item
      info={
        Object {
          "budget": 175000000,
          "genres": Array [
            "Adventure",
            "Science Fiction",
            "Action",
          ],
          "id": 333339,
          "overview": "When the creator of a popular video game system dies, a virtual contest is created to compete for his fortune.",
          "poster_path": "https://image.tmdb.org/t/p/w500/pU1ULUq8D3iRxl1fdX2lZIzdHuI.jpg",
          "release_date": "2018-03-28",
          "revenue": 0,
          "runtime": 140,
          "tagline": "A better reality awaits.",
          "title": "Ready Player One",
          "vote_average": 8.1,
          "vote_count": 617,
        }
      }
      key="333339"
      searchInFilmModeByGenre={[MockFunction]}
      setFilmMode={[MockFunction]}
    />
  </div>
</React.Fragment>
`;

exports[`<Results /> Rendering should render the component without results 1`] = `
<div
  className="results"
>
  <NoResults />
</div>
`;
